### PR TITLE
Create postgresql_slot.py

### PR DESCRIPTION
**Pull request by jbscalia**
This is a new module for managing postgresql slots used in replication. I created this module as part of a replication role. Previously I had to use a shell script which was not idempotent.

##### SUMMARY
PostgreSQL uses slots for both physical and logical replication configurations, and prior to this module, no module performed management of slots.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
postgresql_slot

##### ANSIBLE VERSION
```
ansible 2.6.0 (postgresql_slot_branch e5b3416605) last updated 2018/05/07 11:10:54 (GMT -400)
  config file = /home/419635/.ansible.cfg
  configured module search path = [u'/home/419635/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/419635/ansible/lib/ansible
  executable location = /home/419635/ansible/bin/ansible
  python version = 2.7.5 (default, May  3 2017, 07:55:04) [GCC 4.8.5 20150623 (Red Hat 4.8.5-14)]
```

### CODE REVIEW

**Comment by jbscalia** [comment]
I do not understand why so many documentation issues are listed. I can run ansible-test  on my development server, and it only produces:

ansible-test sanity --test ansible-doc --python 2.7 postgresql_slot
Sanity check using ansible-doc with Python 2.7

**Comment by jbscalia** [comment]
shipit

**Review by Andersson007 — CHANGES_REQUESTED** [review]
@jbscalia , thank you for this good PR,

1. I think "short_description" should be swapped with "description".

2. I suggest to use "Adds or removes.." in descriptions (with "s" at the end of the words here like into other modules in PostgreSQL block), but it's not really important.

3. **db** - Add info that this is also the database to connect to. 

4.  "Name of the slot" Maybe shortly "Slot name"?

5. Maybe better to make state=absent and (decoder or type) as mutually exclusive?

6. Specify PostgreSQL version since Replication slots have been added in the "notes:" section.

7. Better to add a link to the official documentation about creating replication slots.

I want to try it if you make changes.

**Comment by jbscalia** [comment]
Sure, I’ll take a look at these changes later today. Thanks for the feedback.

**Comment by jbscalia** [comment]
my version runs correctly on my server, so I', hoping the Ansibot finds nothing wrong, but it seems awfully fussy.

File: lib/ansible/modules/database/postgresql/postgresql_slot.py
Comment (Andersson007) [inline lib/ansible/modules/database/postgresql/postgresql_slot.py]: change 'string' to 'str'
this recent change in Ansible ('str', 'bool', etc.)

**Review by Andersson007 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/database/postgresql/postgresql_slot.py
Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_slot.py]: and remove a blank lines near line 226

**Comment by Andersson007** [comment]
don't panic ) just 10 minutes

**Comment by Andersson007** [comment]
Did you use this check in your repo dir?
```
ansible# source ./hacking/env-setup
ansible# ansible-test sanity --python 2.7  --base-branch devel lib/ansible/modules/database/postgresql/postgresql_slot.py
```
I'm going to send you proper file in 10 minutes.

**Comment by Andersson007** [comment]
@jbscalia , I just sent fixed file to your mail, please, swap it and push.
On Monday I test this functionality.

**Comment by jbscalia** [comment]
I received your file, but I can only access it from work on my iPhone. The blasted government will not permit me to connect to gmail here. Will swap out the old file for yours once I get home.

File: lib/ansible/modules/database/postgresql/postgresql_slot.py
Comment (kostiantyn-nemchenko) [inline lib/ansible/modules/database/postgresql/postgresql_slot.py]: Would it be better to stick with the naming convention used in Postgres [docs](https://www.postgresql.org/docs/current/view-pg-replication-slots.html) and use `plugin` or `output_plugin` instead of `decoder`

File: lib/ansible/modules/database/postgresql/postgresql_slot.py
Comment (kostiantyn-nemchenko) [inline lib/ansible/modules/database/postgresql/postgresql_slot.py]: It's generally not a good idea to mask [built-in](https://docs.python.org/2/library/functions.html#type) functions.
One more reason to pick `slot_type` as per the [docs](https://www.postgresql.org/docs/current/view-pg-replication-slots.html).

File: lib/ansible/modules/database/postgresql/postgresql_slot.py
Comment (kostiantyn-nemchenko) [inline lib/ansible/modules/database/postgresql/postgresql_slot.py]: Since `physical` is default I suggest leaving this empty

File: lib/ansible/modules/database/postgresql/postgresql_slot.py
Comment (kostiantyn-nemchenko) [inline lib/ansible/modules/database/postgresql/postgresql_slot.py:96]: Hardcoded major versions should be replaced by "current" to always provide up to date pages for end users

**Comment by dagwieers** [comment]
@jbscalia What exactly is it you are trying to do ? If you make changes to commits you already pushed, then git will complain and the only thing you can do is to force-push those changes. But I am not sure this is what you wanted to do.

Also you can save yourself the time going back-and-forth with CI testing byt running `ansible-test sanity <path/to/your/module>` and fix everything before pushing it to GitHub. We are doing automated testing that goes beyond simply running it.

**Comment by jbscalia** [comment]
Dag,

I was trying to replace the file with a corrected version given to me by Andresson, and I had logged in and tried to set edit. I was pretty certain that the replacement file would pass all tests, but github kept crashing when I hit edit. Now, none of this was happening when I was editing this at my office, only here at home, and since I’d been working on this for about 10 hours, I was just more than a bit frustrated.

**Comment by dagwieers** [comment]
@jbscalia I think it is easier to learn to use git command-line (or using an editor with a Git interface). Being able to use `ansible-test` is pretty much a requirement unless you fancy getting corrected by @ansibot ;-)

PS You can also give @Andersson007 write-access to your branch so he can continue/finish with this PR.

**Comment by jbscalia** [comment]
Hmm, i’ve never tried using a command line for github, although I’m quite familiar with git. I’ll look at it.

**Comment by jbscalia** [comment]
I've checked and do not have approval for downloading this repo so I could do the command line operations. Sigh... How would I go about granting Andersson007 write access so he could do the adjustments?

**Comment by dagwieers** [comment]
@jbscalia It is described here: https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

It basically is adding him as a collaborator to your forked repository.

**Comment by jbscalia** [comment]
It does appear to me like Andressen007 and other maintainers can make changes to this pull request. At least I'm seeing the checkbox is in fact, checked.

**Review by kostiantyn-nemchenko — CHANGES_REQUESTED** [review]
Sorry, didn't notice that earlier, but could you please add support of using a [unix socket](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L45), [ssl-related options](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L49) and recently added [session role](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L69)?

**Comment by jbscalia** [comment]
> We also need integration tests (You should write it).

Are there some examples of these tests and where should I place them? This is a fairly new experience for me.

**Comment by Andersson007** [comment]
@jbscalia you can look at examples of postgresql integration tests here ansible/test/integration/targets/postgresql/defaults/

https://github.com/ansible/community/issues/435 - There are some useful (and necessary) information from @dagwieers about integration tests.
See also official documentation:
https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html

If you're using a linux machine for tests (you should do it), you can test it before pushing your commit:
1. Prepare your system:
```
# pip install argcomplete
# pip3 install argcomplete
# yum install docker
# systemctl enable docker && systemctl start docker
```
From ansible src root directory:
```
# source ./hacking/env-setup
```

2. Run test in the docker container:
```
test/runner/ansible-test integration --docker-no-pull --docker fedora28 -v postgresql >> ~/test.log
```
3. Parse log file for tests result, e.g.
```
less ~/test.log
```
Use searching in less, e.g.:
```
/postgresql_slot
```
Maybe there is easier way, I don't know, but seems this method works well.

**Comment by jbscalia** [comment]
> Sorry, didn't notice that earlier, but could you please add support of using a [unix socket](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L45), [ssl-related options](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L49) and recently added [session role](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/database/postgresql/postgresql_ext.py#L69)?

I'm not seeing anything the current PostgreSQL documents regarding what you've suggested for this slot module. I am seeing a new parameter though that this does not cover: immediately_reserve.  So, I guess I should be adding that.

**Comment by kostiantyn-nemchenko** [comment]
> I'm not seeing anything the current PostgreSQL documents regarding what you've suggested for this slot module.

@jbscalia Have you checked the links I posted above?
I mean `login_unix_socket`, `ssl_mode`, `ssl_rootcert` and `session_role` options which should be present in all Postgres-related modules for Ansible.

**Comment by jbscalia** [comment]
>> I mean `login_unix_socket`, `ssl_mode`, `ssl_rootcert` and `session_role` options which should be present in all Postgres-related modules for Ansible.

OK, I see what you mean, sort of. I've looked at some other postgresql modules and I see everything except session_role. Is that parameter new with Ansible 2.8? It's not present in any other module on my host server.

**Comment by kostiantyn-nemchenko** [comment]
> Is that parameter new with Ansible 2.8?

Yes, it has been added recently by https://github.com/ansible/ansible/pull/43650 and will be in the next release.

**Comment by jbscalia** [comment]
I've gotten the code adjusted on my server so it includes Kostiantyn's parameters and I've added both immediately_reserve as well as temporary, but I'm wondering if temporary is really necessary. Mind you it appears that I must make the creation call with either 1 argument or all 3 arguments, 2 arguments does not work, but since temporary says the slot will only exist for this session, I think it would disappear after the play is completed, which wouldn't be of any real use.

BTW, I'm having some problems getting the integration tests running as the docker installation is failing due to a dependency check with yum. Still trying to work that out.

**Comment by jbscalia** [comment]
Yes, I was correct about the temporary parameter. Ran a test with that set to True and there is no slot on the server afterwards. So, unless someone objects vociferously I'm going to hard code that to False and not allow it be settable.

**Comment by jctanner** [comment]
!waffling ci_verified

**Comment by kostiantyn-nemchenko** [comment]
@jbscalia Please, take a look at [this](https://github.com/jbscalia/ansible/pull/1) PR.

**Comment by Andersson007** [comment]
@jbscalia hello,
are you going to fix this module? If yes, cool, I'll make a full review after that.
If you don't have enough time or for other reasons, you don't mind if I (or anybody else) completely rewrite this module (keeping you as an co-author, of course) ?

**Comment by Andersson007** [comment]
!needs_info

**Comment by Akasurde** [comment]
superseded by #55079

**Comment by Andersson007** [comment]
@Akasurde  , thank you

**Pull request closed without merging**
